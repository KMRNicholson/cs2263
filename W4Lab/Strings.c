#include <string.h>
#include <stdlib.h>
#include <stdio.h>

// a cover function for malloc()
// malloc and return memory for a string of stringsize characters
// return (char*)NULL on failure
char* mallocString(int stringsize){
    char* newString = (char*)malloc(stringsize);
    return newString;
}

// just a cover function for free()
void freeString(char* s){
    free(s);
}

// create a duplicate string of s
// return it
// return (char*)NULL on failure
// should call mallocString(), and then strcpy()
char* duplicateString(char* s){
    int copySize;
    int stringSize = strlen(s) + 1;
    char* sptr = mallocString(stringSize);
    if(sptr != (char*)NULL){
        sptr = strncpy(sptr, s, stringSize);
    }
    return sptr;
}

// writes a string to a file, preceded by it's length
// return 0 on success
// return 1 on failure
int fputString(FILE* pFOut, char* s){
    int stringLength = strlen(s);
    int write = fprintf(pFOut, "%d%s", stringLength, s);

    if(write < 0){
        return EXIT_FAILURE;
    }
    
    return EXIT_SUCCESS;
}

// reads a string which is preceded by its length from a file.
// return pointer to the allocated string in the heap
// return (char*)NULL on failure
char* fgetString(FILE* pFIn){
    int stringLength;
    char* string;

    fscanf(pFIn, "%d", &stringLength); // Scan string length

    string = mallocString(stringLength); // Allocate memory
    if(string != (char*)NULL){
        fgets(string, stringLength, pFIn); // Get the rest of the line
    }

    return string;
}