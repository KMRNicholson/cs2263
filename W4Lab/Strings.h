#include <stdio.h>

#ifndef STRINGS_H
#define STRINGS_H

// a cover function for malloc()
// malloc and return memory for a string of stringsize characters
// return (char*)NULL on failure
char* mallocString(int stringsize);

// just a cover function for free()
void freeString(char* s);

// create a duplicate string of s
// return it
// return (char*)NULL on failure
// should call mallocString(), and then strcpy()
char* duplicateString(char* s);

// writes a string to a file, preceded by it's length
// return 0 on success
// return 1 on failure
int fputString(FILE* pFOut, char* s);

// reads a string from file.
// return pointer to the allocated string in the heap
// return (char*)NULL on failure
char* fgetString(FILE* pFIn);

#endif