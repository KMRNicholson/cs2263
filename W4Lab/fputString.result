24This is string number 1
26This is the second string
