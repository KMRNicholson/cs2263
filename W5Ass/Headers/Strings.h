#include <stdio.h>

#ifndef STRINGS_H
#define STRINGS_H

typedef char* String;

// a cover function for malloc()
// malloc and return memory for a string of stringsize characters
// return (char*)NULL on failure
String mallocString(int stringsize);

// just a cover function for free()
void freeString(String s);

// create a duplicate string of s
// return it
// return (char*)NULL on failure
// should call mallocString(), and then strcpy()
String duplicateString(String s);

// create a duplicate list of strings
// return it
// return (char*)NULL on failure
// should call mallocString(), and then strcpy()
String* duplicateStringList(String* s, int listSize);

// writes a string to a file, preceded by it's length
// return 0 on success
// return 1 on failure
int fputString(FILE* pFOut, String s);

// reads a string from file.
// return pointer to the allocated string in the heap
// return (char*)NULL on failure
String fgetString(FILE* pFIn);

// reads a string from stdin
// return pointer to the allocated string
String getString(String input, int n);

// Takes two null pointers to two strings and compares them.
// usable in the qsort() function
// returns 0 if identical
int stringCompare(const void* str1, const void* str2);

void stringToString(String string);

#endif