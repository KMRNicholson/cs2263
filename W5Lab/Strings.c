#include <string.h>
#include <stdlib.h>
#include <stdio.h>

typedef char* String;

// a cover function for malloc()
// malloc and return memory for a string of stringsize characters
// return (String)NULL on failure
String mallocString(int stringsize){
    String newString = (String)malloc(stringsize);
    return newString;
}

// just a cover function for free()
void freeString(String s){
    free(s);
}

// create a duplicate string of s
// return it
// return (String)NULL on failure
// should call mallocString(), and then strcpy()
String duplicateString(String s){
    int stringSize = strlen(s)+1;
    String sptr = mallocString(stringSize);
    if(sptr != (String)NULL){
        sptr = strncpy(sptr, s, stringSize-1);
	    sptr[stringSize-1] = '\0';
    }
    return sptr;
}

// create a duplicate list of strings
// return it
// return (String)NULL on failure
String* duplicateStringList(String* s, int listSize){
    int i = 0;
    String* dupList = (String*)malloc(listSize*sizeof(String));
    if(dupList != (String*)NULL){
        while(i < listSize){
            dupList[i] = duplicateString(s[i+1]);
            i++;
        }
    }
    return dupList;
}

// writes a string to a file, preceded by it's length
// return 0 on success
// return 1 on failure
int fputString(FILE* pFOut, String s){
    int stringLength = strlen(s);
    int write = fprintf(pFOut, "%d%s", stringLength, s);

    if(write < 0){
        return EXIT_FAILURE;
    }
    
    return EXIT_SUCCESS;
}

// reads a string which is preceded by its length from a file.
// return pointer to the allocated string in the heap
// return (String)NULL on failure
String fgetString(FILE* pFIn){
    int stringLength;
    String string;

    fscanf(pFIn, "%d", &stringLength); // Scan string length

    string = mallocString(stringLength); // Allocate memory
    if(string != (String)NULL){
        fgets(string, stringLength+1, pFIn); // Get the rest of the line
    }

    return string;
}

// reads a string from stdin
// return pointer to the allocated string
String getString(String input, int n){
    fgets(input, n-1, stdin);
    return input;
}

// Takes two null pointers to two strings and compares them.
// usable in the qsort() function
// returns 0 if identical
int stringCompare(void* str1, void* str2){
    const String* ptr1 = (String*)str1;
    const String* ptr2 = (String*)str2;
    return strcmp(*ptr1, *ptr2);
}