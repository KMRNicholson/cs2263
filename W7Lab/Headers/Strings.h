/* 
    Strings.h

    Description:
    Header file which contains function signatures for a char*.

    Author: 
    Kohdy Nicholson
    
    Date: 
    2020-06-03
*/

#include <stdio.h>

#ifndef STRINGS_H
#define STRINGS_H

// Typedef
typedef char* String;

// mallocs a string
String mallocString(int stringsize);
// free's a string
void freeString(String s);
// deep copies a string
String duplicateString(String s);
// duplicate a list of strings
String* duplicateStringList(String* s, int listSize);
// writes a string to a file
int fputString(FILE* pFOut, String s);
// reads a string from file
String fgetString(FILE* pFIn);
// reads a string from stdin
String getString();
// Takes two null pointers to two strings and compares them.
int stringCompare(const void* str1, const void* str2);
// prints a string
void stringToString(String string);

#endif