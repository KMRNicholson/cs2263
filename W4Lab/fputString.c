/* 
    fputString.c

    Description:
    Program that utilizes various file operations with strings.
    
    Author: 
    Kohdy Nicholson
    
    Date: 
    2020-05-26
*/

#include "Strings.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char** argv){
    int n = 100;
    int i = 0;
    char stdString[n];
    
    FILE* file = fopen(argv[1], "w");
    if(file == (FILE*)NULL){
        printf("Failed to open file. Exiting..");
        return EXIT_FAILURE;
    }

    while(i < 2){
        fgets(stdString, n-1, stdin);
        fputString(file, stdString);
        i++;
    }

    return EXIT_SUCCESS;
}