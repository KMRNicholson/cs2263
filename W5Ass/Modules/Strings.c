#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include "../Headers/Strings.h"

// a cover function for malloc()
// malloc and return memory for a string of stringsize characters
// return (String)NULL on failure
String mallocString(int stringsize){
    String newString = (String)malloc(stringsize);
    return newString;
}

// just a cover function for free()
void freeString(String s){
    free(s);
}

// create a duplicate string of s
// return it
// return (String)NULL on failure
// should call mallocString(), and then strcpy()
String duplicateString(String s){
    int stringSize = strlen(s);
    String sptr = mallocString(stringSize+1);
    if(sptr != (String)NULL){
        sptr = strncpy(sptr, s, stringSize);
	    sptr[stringSize] = '\0';
    }
    return sptr;
}

// create a duplicate list of strings
// return it
// return (String)NULL on failure
String* duplicateStringList(String* s, int listSize){
    int i = 0;
    String* dupList = (String*)malloc(listSize*sizeof(String));
    if(dupList != (String*)NULL){
        while(i < listSize){
            dupList[i] = duplicateString(s[i+1]);
            i++;
        }
    }
    return dupList;
}

// writes a string to a file, preceded by it's length
// return 0 on success
// return 1 on failure
int fputString(FILE* pFOut, String s){
    int stringLength = strlen(s);
    int write = fprintf(pFOut, "%d%s", stringLength, s);

    if(write < 0){
        return EXIT_FAILURE;
    }
    
    return EXIT_SUCCESS;
}

// reads a string which is preceded by its length from a file.
// return pointer to the allocated string in the heap
// return (String)NULL on failure
String fgetString(FILE* pFIn){
    char fileInput[255];
    String string;
    fscanf(pFIn, "%[^\n]", fileInput);
    string = duplicateString(fileInput);
    return string;
}

// reads a string from stdin
// return pointer to the allocated string
String getString(){
    char input[255];
    String string;
    scanf("%[^\n]", input);
    string = duplicateString(input);
    return string;
}

// Takes two null pointers to two strings and compares them.
// usable in the qsort() function
// returns 0 if identical
int stringCompare(const void* str1, const void* str2){
    const String* ptr1 = (String*)str1;
    const String* ptr2 = (String*)str2;
    return strcmp(*ptr1, *ptr2);
}

void stringToString(String string){
    printf("%s", string);
}