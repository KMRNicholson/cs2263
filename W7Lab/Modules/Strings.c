/* 
    Strings.c

    Description:
    Code file which contains function for a char*.

    Author: 
    Kohdy Nicholson
    
    Date: 
    2020-06-08
*/

#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include "../Headers/Strings.h"

// mallocs a string
String mallocString(int stringsize){
    String newString = (String)malloc(stringsize);
    return newString;
}

// free's a string
void freeString(String s){
    free(s);
}

// deep copies a string
String duplicateString(String s){
    int stringSize = strlen(s);
    String sptr = mallocString(stringSize+1);
    if(sptr != (String)NULL){
        sptr = strncpy(sptr, s, stringSize);
	    sptr[stringSize] = '\0';
    }
    return sptr;
}

String* duplicateStringList(String* s, int listSize){
    int i = 0;
    String* dupList = (String*)malloc(listSize*sizeof(String));
    if(dupList != (String*)NULL){
        while(i < listSize){
            dupList[i] = duplicateString(s[i+1]);
            i++;
        }
    }
    return dupList;
}

// writes a string to a file
int fputString(FILE* pFOut, String s){
    int stringLength = strlen(s);
    int write = fprintf(pFOut, "%d%s", stringLength, s);

    if(write < 0){
        return EXIT_FAILURE;
    }
    
    return EXIT_SUCCESS;
}

// reads a string from file
String fgetString(FILE* pFIn){
    char fileInput[255];
    String string;
    fscanf(pFIn, "%[^\n]", fileInput);
    string = duplicateString(fileInput);
    return string;
}

// reads a string from stdin
String getString(){
    char input[255];
    String string;
    scanf("%[^\n]", input);
    string = duplicateString(input);
    return string;
}

// Takes two null pointers to two strings and compares them.
int stringCompare(const void* str1, const void* str2){
    const String* ptr1 = (String*)str1;
    const String* ptr2 = (String*)str2;
    return strcmp(*ptr1, *ptr2);
}

// prints a string
void stringToString(String string){
    printf("%s", string);
}