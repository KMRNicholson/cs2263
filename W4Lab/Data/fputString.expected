24This is string number 1
25This is the second string